hack/generator/spec/versioning.md

```diff
@@ -0,0 +1,201 @@
+| NodeTypes                         | []NodeTypes                      | []NodeTypeDescription              | Change of Type   | Property Bag |
+| ReliabilityLevel                  | *Level                           | *ClusterPropertiesReliabilityLevel | Change of Type   | Property Bag |
+| ReverseProxyCertificate           |                                  | *CertificateDescription            | New Property     | Skip         |
+| UpgradeDescription                | *PaasClusterUpgradePolicy        | *ClusterUpgradePolicy              | Change of Type   | Property Bag |
+| UpgradeMode                       |                                  | *ClusterPropertiesUpgradeMode      | New Property     | Skip         |
+| VmImage                           | *string                          | *string                            | None             | Copy value   |
+
+Of the 15 properties listed:
+
+* 8 are unchanged (same name and type); these values are simply copied across.  
+  (AzureActiveDirectory, Certificate, ClientCertificateCommonNames, ClientCertificateThumbprints, DiagnosticsStorageAccountConfig, FabricSettings, ManagementEndpoint, and VmImage)
+* 3 new properties are introduced; these are ignored by the conversion.  
+  (ClusterCodeVersion, ReverseProxyCertificate, and UpgradeMode)
+* 1 property was dropped; it will be stored in the name/value property bag.  
+  (HttpApplicationGatewayCertificate)
+* 3 properties had their types changed; these will also be serialized into the name/value property bag.  
+  (NodeTypes, ReliabilityLevel, and UpgradeDescription)
+
+
+
+
+## Testing
+
+It's vital that we are able to correctly round trip every spoke version to the hub version and back again, so we will generate unit tests to ensure that this works properly. This will help to enusre a base level of compliance, that information is not lost through serialization.
+
+
+
+## Alternative Solutions
+
+AKA the road not travelled
+
+### Alternative: Superset storage schema
+
+The "v1" storage version of each supported resource type will be created by merging all of the fields of all the distinct versions of the resource type, creating a *superset* type that includes every property declared across every version of the API.
+
+To maintain backward compatibility as Azure APIs evolve over time, we will include properties across all versions of the API, even for versions we are not currently generating as output. This ensures that properties in use by older APIs are still present and available for forward conversion to newer APIs, even as those older APIs age out of use.
+
+#### Limitation: Old properties aging out
+
+When a changed property ages out and the schema is no longer available, the code generator will be unaware of its existence. Therefore, it will be omitted from our generated storage schema resulting in a breaking change for any resource already persisted.
```

Comment:
This would only happen if we're not supporting that older version, right? So I think it comes down to the decision on when to stop supporting an older version, recognizing that anyone using that version will likely be broken?

Reply:
There are two reasons why an old API version might drop out:

* By our policy, that we decide we're only supporting the 3 latest GA versions. We could decide to support _all_ of them, but that makes our support load greater.

* By ARM policy, when the older ARM JSON schema is no longer available because ARM are not publishing it. In this case we don't have a choice.